### ISSUE TITLE
Featurerequest: Formatting of zipcodes

### ISSUE THREAD

**Issue by aiomaster** [dgilperez/validates_zipcode#43]
Hi,
Besides validation I want to format zipcodes according to country specific rules.
I have seen, that there is a gem https://github.com/globaldev/going_postal that tries to do this, but it seems unmaintained.
Do you think formatting could be a part of this gem or do you only want the validation thing covered?
Thanks.

**Comment by dgilperez** [comment]
That is interesting @aiomaster, I haven't thought about it! The use case would be to store the codes in any accepted format, and then be able to present them in an harmonized way? 

I think it can be an interesting addition to the gem, kind of closing the gap ... but it also does look like a significant amount of work: it'd probably require to maintain a second list of transformers to play nice with validations. 

I reckon something like would do: 

```
ValidatesZipcode.new("sl41eg", :gb).to_s  #=> "SL4 1EG"
ValidatesZipcode.new("SL4 1EG", :gb).to_s  #=> "SL4 1EG"
ValidatesZipcode.new("sl41egxxxxx", :gb).to_s  #=> raises InvalidError
```

Would you be interested in working in such a PR?

**Comment by aiomaster** [comment]
Yeah, I think I can start a PR and build a formatter where country specific formatting rules can be implemented.
But I can't provide all the rules by myself, just for a few countries to start with.
We should discuss some basic behaviour. Your examples make semantically sense to me, but have a strange api. I think you mean something like that:
```ruby
ValidatesZipcode::Zipcode.new(zipcode: "sl41eg", country_alpha2: :gb).to_s  #=> "SL4 1EG"
ValidatesZipcode::Zipcode.new(zipcode: "SL4 1EG", country_alpha2: country_alpha2: :gb).to_s  #=> "SL4 1EG"
ValidatesZipcode::Zipcode.new(zipcode: "sl41egxxxxx", country_alpha2: :gb).to_s  #=> raises InvalidError
```
or maybe that:
```ruby
ValidatesZipcode.format("sl41eg", :gb)  #=> "SL4 1EG"
ValidatesZipcode.format("SL4 1EG", :gb)  #=> "SL4 1EG"
ValidatesZipcode.format("sl41egxxxxx", :gb)  #=> raises InvalidError
```
but what about
```ruby
ValidatesZipcode.format("  sl41eg", :gb)
```
Should it just work and strip the leading whitespace or should it raise an error?
I think about some basic sanitizing before doing the validation like stripping whitespaces or upcasing anything. Another idea is to make an extra option for that (defaulting to true):
```ruby
ValidatesZipcode.format("  sl41eg", :gb) #=> "SL4 1EG"
ValidatesZipcode.format("  sl41eg", :gb, sanitize: false) #=> raises InvalidError
ValidatesZipcode.format("sl41eg", :gb, sanitize: false) #=> "SL4 1EG"
```

What do you think? Does that make sense to you?

**Comment by dgilperez** [comment]
Yeah, I think you're right. `ValidatesZipcode.format("  sl41eg", :gb)` looks consistent with current `ValidatesZipcode.valid?` API. We could use keyword arguments as you point out, but for only 2 attrs I think it's clear enough as it is. It's also more consistent again with `.valid?`, that doesn't use keyword args. 

About the stripping, I think it's going a bit too ahead and it may be unconsistent with current validator: `"  sl41eg"` is probably not a valid zipcode in most regexes ... but would be formatted? I can also see the argument in favor of that: strip, validate, and then handle case and dashes and things that are optional, to choose the canonical version. 

About the upcasing, I'm not sure that upcase is "canonical" or "good" zipcode for every country. It probably is, just not sure (don't know all formats by heart ofc!).

It looks to me we'd need to have some sensible default and then go country by country. We'd model the formatting as a chain of transformations, something like this: 

```
FORMATS = {
  gb: ->(code) { default_format(code).upcase } 
  xx: ->(code) { default_format(code).upcase.tr(' ', '-') }
  ... 
  ... other particular cases
}
def self.format(zipcode, country_alpha2)
  FORMATS[] || default_format(zipcode)
end

def default_format(zipcode)
  zipcode.to_s.strip
end
```

Does this makes sense?

**Comment by aiomaster** [comment]
I think it would be better to have a method for every country, cause that way it is easier to do the same formatting for different countries.
I think of something like this:
```ruby
module ValidatesZipcode
  module Formatter
    class << self
      def format(zipcode, country_alpha2)
        method = "format_#{country_alpha2}"
        respond_to?(method) ? send(method, zipcode) : format_default(zipcode)
      end

      def format_default(zipcode)
         zipcode.to_s.strip
      end

      def format_cz(zipcode)
        zipcode.to_s.scan(/\d/).insert(3, ' ').join
      end

      alias format_sk format_cz
    end
  end
end
```

So it would be more easy to say, that the formatting for cz and sk is the same. If you are fine with that I will start a PR with some (but not all) countries.

**Comment by dgilperez** [comment]
Closed by #44
